File: sdk/typescript/src/wrap.ts
/**
 * GliaxinWrapper — one-line turn-aware memory for any LLM.
 *
 * @example
 * import { Gliaxin } from 'gliaxin'
 *
 * const g = new Gliaxin('glx_YOUR_KEY')
 *
 * // Provide YOUR LLM as any async function: (messages: Message[]) => Promise<string>
 * async function myLlm(messages: { role: string; content: string }[]) {
 *   const res = await openai.chat.completions.create({ model: 'gpt-4o', messages })
 *   return res.choices[0].message.content!
 * }
 *
 * // agent_name is required — the wrapper auto-registers it on first use
 * const client = g.wrap(myLlm, { agentName: 'support-bot' })
 *
 * // Memory is automatic. Both user and assistant turns are saved.
 * const reply = await client.chat({ userId: 'user_123', message: 'What do I like?' })
 */

import type { Memory } from './types.js'

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant' | string
  content: string
}

export type LlmFn = (messages: ChatMessage[]) => Promise<string>

export interface WrapOptions {
  /** Required. Human-readable name for the agent. Auto-registered on first call. */
  agentName: string
  /** Memory scope for saved turns. Default: "agent" (private). */
  scope?: 'agent' | 'project'
  /** Max memories injected per call. Default: 10 */
  contextLimit?: number
  /** Auto-register the agent on first call. Default: true */
  autoRegister?: boolean
  /** Auto-save the full turn after each call. Default: true */
  autoSave?: boolean
  /** Header line prepended before the memory block in the system prompt. */
  systemPrefix?: string
}

export interface ChatOptions {
  /** Your end-user's ID. */
  userId: string
  /** The user's message text. */
  message: string
  /** Optional prior turns for multi-turn context. */
  history?: ChatMessage[]
  /** Optional base system prompt. Memory context is appended to it. */
  system?: string
}

export class GliaxinWrapper {
  private readonly _g: any
  private readonly _llm: LlmFn
  private readonly _agentName: string
  private readonly _scope: 'agent' | 'project'
  private readonly _contextLimit: number
  private readonly _autoRegister: boolean
  private readonly _autoSave: boolean
  private readonly _systemPrefix: string
  private _agentId: string | undefined

  constructor(gliaxin: any, llm: LlmFn, options: WrapOptions) {
    this._g             = gliaxin
    this._llm           = llm
    this._agentName     = options.agentName
    this._scope         = options.scope         ?? 'agent'
    this._contextLimit  = options.contextLimit  ?? 10
    this._autoRegister  = options.autoRegister  ?? true
    this._autoSave      = options.autoSave      ?? true
    this._systemPrefix  = options.systemPrefix  ?? 'You have access to the following memories about this user:'
  }

  private async _ensureRegistered(): Promise<string> {
    if (this._agentId === undefined) {
      const result = await this._g.agent.register(this._agentName)
      this._agentId = result.agent_id
    }
    return this._agentId!
  }

  /**
   * Send a message with automatic memory injection and full-turn auto-save.
   *
   * @example
   * const reply = await client.chat({ userId: 'user_123', message: 'What do I like?' })
   */
  async chat(opts: ChatOptions): Promise<string> {
    const { userId, message, history, system } = opts

    const agentId = this._autoRegister ? await this._ensureRegistered() : this._agentId

    const memories: Memory[] = await this._g.memory.search(userId, message, {
      limit: this._contextLimit,
      agentId,
    })

    const messages = this._buildMessages(message, memories, history, system)

    const reply = await this._llm(messages)

    if (this._autoSave && agentId) {
      await this._saveTurn(userId, agentId, message, reply)
    }

    return reply
  }

  private _buildMessages(
    message: string,
    memories: Memory[],
    history?: ChatMessage[],
    system?: string,
  ): ChatMessage[] {
    const messages: ChatMessage[] = []
    const systemParts: string[] = []

    if (system) systemParts.push(system)

    if (memories.length > 0) {
      const lines = this._systemPrefix ? [this._systemPrefix] : []
      for (const m of memories) {
        lines.push(`- [${m.category}] ${m.content}`)
      }
      systemParts.push(lines.join('\n'))
    }

    if (systemParts.length > 0) {
      messages.push({ role: 'system', content: systemParts.join('\n\n') })
    }

    if (history) messages.push(...history)

    messages.push({ role: 'user', content: message })
    return messages
  }

  private async _saveTurn(userId: string, agentId: string, userMsg: string, assistantReply: string): Promise<void> {
    try {
      await this._g.memory.addTurn(
        userId,
        agentId,
        [
          { role: 'user',      content: userMsg },
          { role: 'assistant', content: assistantReply },
        ],
        { scope: this._scope },
      )
    } catch {
      // save errors never break the caller's response flow
    }
  }
}

File: sdk/go/wrap.go
package gliaxin

import (
	"context"
	"fmt"
	"strings"
	"sync"
)

// LLMFunc is any function that accepts a list of messages and returns a reply.
// Works with OpenAI, Anthropic, Gemini, Ollama, LiteLLM — anything that
// takes a []Message and returns (string, error).
type LLMFunc func(ctx context.Context, messages []Message) (string, error)

// WrapOptions configures a Wrapper returned by Client.Wrap.
type WrapOptions struct {
	// AgentName is required. Human-readable name for the agent.
	// Auto-registered on first Chat call (idempotent).
	AgentName string
	// Scope controls memory visibility for saved turns.
	// "agent" = private to this agent (default).
	// "project" = shared across all agents in the project.
	Scope string
	// ContextLimit is the max memories injected per call (default 10).
	ContextLimit int
	// DisableAutoRegister skips agent registration on first call.
	// By default auto-register is ON.
	DisableAutoRegister bool
	// DisableAutoSave disables automatic saving of the full turn after each call.
	// By default auto-save is ON.
	DisableAutoSave bool
	// SystemPrefix is prepended before the memory block in the system prompt.
	// Defaults to "You have access to the following memories about this user:".
	// Set to a single space " " to suppress the header while keeping memories.
	SystemPrefix string
}

// ChatOptions are per-call options for Wrapper.Chat.
type ChatOptions struct {
	// History is optional prior conversation turns.
	History []Message
	// System is an optional base system prompt.
	// Retrieved memories are appended to it automatically.
	System string
}

// Wrapper wraps any LLMFunc with automatic turn-based memory.
// Create one via Client.Wrap.
type Wrapper struct {
	g       *Client
	llm     LLMFunc
	opts    WrapOptions
	agentID string
	mu      sync.Mutex
}

// Chat sends a message with automatic memory injection and full-turn auto-save.
//
// On every call:
//  1. Ensures the agent is registered (idempotent, cached after first call).
//  2. Searches the user's memories using that agent's identity.
//  3. Injects found memories silently into the system prompt.
//  4. Calls your LLM with the enriched message list.
//  5. Saves the full turn (user + assistant) attributed to the agent.
//
// Example:
//
//	reply, err := wrapper.Chat(ctx, "user_123", "What do I like?", gliaxin.ChatOptions{})
func (w *Wrapper) Chat(ctx context.Context, userID, message string, opts ChatOptions) (string, error) {
	agentID, err := w.ensureRegistered(ctx)
	if err != nil {
		return "", fmt.Errorf("gliaxin wrap: agent registration: %w", err)
	}

	limit := w.opts.ContextLimit
	if limit == 0 {
		limit = 10
	}

	scope := w.opts.Scope
	if scope == "" {
		scope = "agent"
	}

	memories, err := w.g.Memory.Search(ctx, userID, message, SearchMemoryOptions{
		Limit:   limit,
		AgentID: agentID,
	})
	if err != nil {
		return "", fmt.Errorf("gliaxin wrap: memory search: %w", err)
	}

	messages := buildMessages(message, memories, opts.History, opts.System, w.opts.SystemPrefix)

	reply, err := w.llm(ctx, messages)
	if err != nil {
		return "", err
	}

	if !w.opts.DisableAutoSave && agentID != "" {
		// Best-effort — save errors never break the caller's response flow.
		_, _ = w.g.Memory.AddTurn(ctx, userID, agentID, []Message{
			{Role: "user",      Content: message},
			{Role: "assistant", Content: reply},
		}, AddTurnOptions{Scope: scope})
	}

	return reply, nil
}

func (w *Wrapper) ensureRegistered(ctx context.Context) (string, error) {
	if w.opts.DisableAutoRegister {
		w.mu.Lock()
		id := w.agentID
		w.mu.Unlock()
		return id, nil
	}
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.agentID != "" {
		return w.agentID, nil
	}
	result, err := w.g.Agent.Register(ctx, w.opts.AgentName)
	if err != nil {
		return "", err
	}
	w.agentID = result.AgentID
	return w.agentID, nil
}

func buildMessages(message string, memories []Memory, history []Message, system, prefix string) []Message {
	var msgs []Message

	var sysParts []string
	if system != "" {
		sysParts = append(sysParts, system)
	}

	if len(memories) > 0 {
		var lines []string
		header := prefix
		if header == "" {
			header = "You have access to the following memories about this user:"
		}
		if strings.TrimSpace(header) != "" {
			lines = append(lines, header)
		}
		for _, mem := range memories {
			lines = append(lines, fmt.Sprintf("- [%s] %s", mem.Category, mem.Content))
		}
		sysParts = append(sysParts, strings.Join(lines, "\n"))
	}

	if len(sysParts) > 0 {
		msgs = append(msgs, Message{
			Role:    "system",
			Content: strings.Join(sysParts, "\n\n"),
		})
	}

	msgs = append(msgs, history...)
	msgs = append(msgs, Message{Role: "user", Content: message})
	return msgs
}
